src/Features/Core/Portable/InvertLogical/AbstractInvertLogicalCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,145 @@
+            var document3 = await InvertOuterExpressionAsync(document2, cancellationToken).ConfigureAwait(false);
+            return document3;
+        }
+
+        private async Task<Document> InvertInnerExpressionAsync(
+            Document document, int position, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // Walk up to the topmost binary of the same type.  When converting || to && (or vice versa)
+            // we want to grab the entire set.  i.e.  `!a && !b && !c` should become `!(a || b || c)` not
+            // `!(a || b) && !c`
+            var binaryExpression = root.FindToken(position).Parent;
+            while (binaryExpression.Parent?.RawKind == binaryExpression.RawKind)
+            {
+                binaryExpression = binaryExpression.Parent;
+            }
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var newBinary = generator.Negate(binaryExpression, semanticModel, cancellationToken);
+
+            return document.WithSyntaxRoot(root.ReplaceNode(
+                binaryExpression,
+                newBinary.WithAdditionalAnnotations(s_annotation)));
+        }
+
+        private async Task<Document> InvertOuterExpressionAsync(
+            Document document, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+
+            var expression = root.GetAnnotatedNodes(s_annotation).Single();
+
+            // Walk up parens and !'s.  That way we don't end up with something like !!.
+            // It also ensures that this refactoring reverses itself when invoked twice.
+            while (syntaxFacts.IsParenthesizedExpression(expression.Parent) ||
+                   syntaxFacts.IsLogicalNotExpression(expression.Parent))
+            {
+                expression = expression.Parent;
+            }
+
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            // Negate the containing binary expr.  Pass the 'negateBinary:false' flag so we don't
+            // just negate the work we're actually doing right now.
+            return document.WithSyntaxRoot(root.ReplaceNode(
+                expression,
+                generator.Negate(expression, semanticModel, negateBinary: false, cancellationToken)));
+        }
+
+        private string GetTitle(TSyntaxKind binaryExprKind)
+            => string.Format(FeaturesResources.Replace_0_with_1,
```

Comment:
>Replace_0_with_1 [](start = 47, length = 16)

I don't like this title, as it could be applicable to all refactorings.
Maybe "Use negated conjunction" or "Use negated disjunction" (depending on the case), or maybe "Apply De Morgan's rule"?

Reply:
> I don't like this title, as it could be applicable to all refactorings.

The title, once subsituted is pretty clear as to waht it's doing.  you'lld see ```Replace && with ||```.  Which i think is fairly clear (and doubly so once you have the preview).

I'm open to alternatives, but i think they need to be clear and not excessively baroque/geeky.  Negated-conjunction and 'De Morgan's Law' are just really "out there" for me.  They are pedantically on point, but i think they would be far less clear for users in general.

Reply:
could also just call it "Invert operation".  But i really think `Replace && with ||` is suitably clear.  And, to be honest, i'm totally open for using Replace_0_with_1 with other refactorings if it makes sense :)

Reply:
Oh, that's not bad. I thought it was `Replace expr1 with expr2`. My bad :-)

---
In reply to: [218197988](https://github.com/dotnet/roslyn/pull/29928#discussion_r218197988) [](ancestors = 218197988)